### DISCUSSION TITLE
Is valgrind reporting a false positive inside aesni_cbc_hmac_sha1_cipher.c?

### DISCUSSION THREAD

**Discussion by slontis** [openssl/openssl#23780] [General Discussion]
Running the following on master

````
LD_LIBRARY_PATH=. valgrind  ./test/sslapitest -test 70 -iter 1 test/certs test/recipes/90-test_sslapi_data/passwd.txt tmp.txt default test/default.cnf test/recipes/90-test_ssl
````
Results in valgrind results of
'Use of uninitialised value of size 8'
When it does 
````
c = p[j];
````

Where
````
unsigned char *p = out + len - 1 - maxpad - SHA_DIGEST_LENGTH;
`````
Has anyone seen this before, and is it expected?

I littered the code with printf statements and the issue propagated further upwards in the file to here
````
            printf("5: %p %p %p %ld %ld %ld\n", in, out, out1, out - out1, len, len1);
            /* figure out payload length */
            pad = out[len - 1];
            printf("PAD1: %d\n", pad); //Valgrind complains about this line 
````

There are all sorts of things going on in this function including the passed in/out parameters being the same, as well as assembler function calls and constant time (inline assembler) calls.

**Comment by nhorman** [comment]
Thats odd.

I can reproduce your errors running your exact command line, but setting LD_LIBRARY_PATH on the valgrind command is not the way I've read that you should run the command (in the event that the environment doesn't get passed on fork).  If I run this command instead:
```
valgrind env LD_LIBRARY_PATH=/home/nhorman/git/worktrees/openssl ./test/sslapitest -test 70 -iter 1 test/certs test/recipes/90-test_sslapi_data/passwd.txt tmp.txt default test/default.cnf test/recipes/90-test_ssl
```

The errors are not reproduced.  Additionally, if I build the library with enable-asan and just run:
```
LD_LIBRARY_PATH=. ./test/sslapitest -test 70 -iter 1 test/certs test/recipes/90-test_sslapi_data/passwd.txt tmp.txt default test/default.cnf test/recipes/90-test_ssl
```
I get no errors.

As to weather or not its a true positive, I can't being to tell

**Comment by mattcaswell** [comment]
I feel fairly sure I have seen this before and we treated it as a false positive. But I don't recall any of the details.

**Comment by kroeckx** [comment]
Note that asan doesn't report uninitialized values. I think you need msan for that.
